Solutions/Exercises/ex5.lean

```diff
@@ -0,0 +1,248 @@
+-- any term of type `α` can be used as a term of type `β`.
+
+-- Conversely, if `x : α` and `x : β` then `α = β`. This is not something we can prove,
+-- it is a syntactical necessity.
+
+-- We can however prove that two types are not equal, if they satisfy different properties.
+lemma ne_types_of_different_property {α β : Type}
+  (p : Type → Prop) (hα : p α) (hβ : ¬ p β) : α ≠ β := by
+rintro rfl; contradiction
+
+-- Now prove that there exists two types that are not equal.
+theorem exists_ne_types : ∃ α β : Type, α ≠ β :=
+⟨ℕ, Empty, λ h => by cases h.mp 1⟩
+
+-- If you get stuck at this point, skip this exercise and come back to it at the end!
+
+section
+
+/-
+  Types can be related in other ways other than being equal. For example, mathlib defines
+  a type `Cardinal` which is a quotient of `Type` by the equivalence relation of having the same
+  cardinality. Two types have the same cardinality if there exists a bijection between them.
+-/
+
+-- `ctrl/cmd + click` through to examine the definitions.
+#check Cardinal
+-- Note how `Nonempty` is used in the definition of `Cardinal`.
+
+-- There are also types that witness when a type is countable or uncountable.
+#check Countable
+#check Uncountable
+-- These are defined as classes, which we haven't seen yet, but you can just think of them as
+-- inductive types with only one constructor for now.
+
+-- Mathlib defines the convenient notation `#α` for the cardinal corresponding to a type `α`.
+-- To access it `#` we need to open the `Cardinal` namespace.
+open Cardinal
+
+variable {α : Type} in
+#check #α
+-- This is the same as `Quotient.mk Cardinal.isEquivalent α`.
+
+-- Now prove that two types where one is countable and one isn't cannot be equal.
+lemma ne_types_of_countable_uncountable {α β : Type}
+  (hα : Countable α) (hβ : Uncountable β) : α ≠ β :=
+ne_types_of_different_property _ hα ((uncountable_iff_not_countable _).mp hβ)
+
+-- Now prove this. (Hint: you don't need to use the previous one.)
+lemma ne_types_of_ne_cardinal {α β : Type} (h : #α ≠ #β) : α ≠ β :=
+by { rintro rfl; contradiction }
+
+-- *Reminder*: if you couldn't find two types that were not equal before, go back to theorem `exists_ne_types` a try it now.
+
+end
+
+
+/-! ## Part 5 : Nat.cast is not surjective
+  For this last question, we will prove that the obvious inclusion `ℕ → ℤ` is not surjective.
+  In Lean, this inclusion is called `Nat.cast`.
+-/
+
+open Function
+
+#check (Nat.cast : ℕ → _)
+-- Cast is defined in greater generality as a function `ℕ → α` for certain types `α`.
+-- This argument is implicit, but we can provide it explicilty using the `(α := ℤ)` syntax.
+#check Nat.cast 0
+-- Note that Lean uses the `↑` notation for `Nat.cast`. We will learn more about this next week.
+
+lemma not_surjective_nat_int : ¬ Surjective (Nat.cast : ℕ → ℤ) := λ h => by
+{ obtain ⟨x, hx⟩ := h (-1)
+  simp only [Int.reduceNeg] at hx }
```

Comment:
Ah, I think you need to update mathlib on your end, and so does everyone else

Reply:
If we want students (and us) to update the dependency versions we should add `lake-manifest.json` to `.gitignore`, and maybe even `lean-toolchain`, shouldn't we?

Reply:
I don't think we should. Because those get updated automatically.

But, anyway, I'm getting errors everywhere after updating now. So let's just stick to not updating it lol